fairlearn/reductions/_moments/cdf_demographic_parity_moment.py

```diff
@@ -0,0 +1,198 @@
+        X_aug['theta'] = pd.Series(theta_list, index=X_aug.index)
+
+        X_aug.index = range(self.n * self.grid_num)
+        A_aug.index = range(self.n * self.grid_num)
+        Y_values.index = range(self.n * self.grid_num)
+
+        def weight_assign(theta, y):
+            return (self.loss.eval([self.next_grid(_theta) for _theta in theta], y) -
+                    self.loss.eval([self.prev_grid(_theta) for _theta in theta], y))
+        W = weight_assign(X_aug['theta'], Y_values)
+        Y_aug = 1*(W < 0)
+        W = abs(W)
+        return X_aug, A_aug, Y_aug, W
+
+    def load_data(self, X, y, **kwargs):
+        super().load_data(X, y, **kwargs)
+        # augment data
+        X_aug, A_aug, Y_aug, W = self.augment_data(self.X, self.tags[_GROUP_ID], self.tags[_LABEL])
+        self.objective.load_augmented_data(X_aug, Y_aug, sensitive_features=A_aug, weights=W)
+
+        # pass augmented data to utility_parity
+        self.utility_parity.load_data(X_aug, Y_aug, event=X_aug["theta"], sensitive_features=A_aug)
+        self.index = self.utility_parity.index
+
+        empirical_y_range = (self.tags[_LABEL].min(), self.tags[_LABEL].max())
+        assert (self.y_range[0] <= empirical_y_range[0])\
+            and (self.y_range[1] >= empirical_y_range[1]),\
+            "The actual range of the label y exceeds that required!"
+
+    def default_objective(self):
+        return self.objective
+
+    def bound(self):
+        return self.utility_parity.bound()
+
+    def project_lambda(self, lambda_vec):
+        return self.utility_parity.project_lambda(lambda_vec)
+
+    def gamma(self, predictor):
+        def classifier(X):
+            x = X.drop(["theta"], axis=1).iloc[:self.n]
+            pred = pd.Series(predictor(x))
+            pred_binary = pd.concat(repeat(pred, self.grid_num))
+            pred_binary.index = range(self.n * self.grid_num)
+            return 1*(pred_binary - X['theta'] >= 0)
+            # made improvements in constrast to the original line
+            # return 1*( predictor(X.drop(['theta'], axis=1)) - X['theta'] >= 0)
+            # noting that,after dropping the theta column, the data are repeated for grid_num times
+        return self.utility_parity.gamma(classifier)
+
+    def signed_weights(self, lambda_vec):
+        return self.utility_parity.signed_weights(lambda_vec)
+
+    def optimal_label(self, signed_weights):
+        """ A heuristic is used to calculate the optimal y, as suggested in the
+        section 4.4 of 'Agarwal et al. (2019) <https://arxiv.org/abs/1905.12843v1>'.
+
+        Parameters
+        ----------
+        signed_weights : pandas.DataFrame
+            signed weights
+        
+        Returns
+        -------
+        pandas.Series
+            The optimal score y for each x before augmentation.
+        """
+        redY = 1 * (signed_weights > 0)
+        redW = signed_weights.abs()
+        redW = self.n * self.grid_num * redW / redW.sum()
+        minimizer = {}
```

Comment:
This name is surprising to me. Is it really minimizing something as `minimizer` would suggest, or is it rather the set of minima? Also, it seems like we're just setting values at the keys 0, 1, 2, 3, ... n-1 which would be more suitable for a list rather than a dict. Is that assumption right?